### ISSUE TITLE
Refineing the WoT.discover() function

### ISSUE THREAD

**Issue by mkovatsc** [w3c/wot-scripting-api#109]
We currently have `Observable<ConsumedThing> discover(optional ThingFilter filter);` to dynamically discover Things. We should spend some time on improvement for the Korea WD release.

The optional filter looks like this:
```
dictionary ThingFilter {
    DiscoveryMethod      method = "any";
    USVString            url;
    USVString            query;
    sequence<Dictionary> constraints;
};
```

The possible methods are:
* "any" does not provide any restriction
* "local" for discovering Things defined in the same device
* "nearby" for discovering Things nearby the device, e.g. by Bluetooth or NFC
* "directory" for discovery based on a service provided by a directory or repository of Things
* "broadcast" for an open ended discovery based on sending a request to a broadcast address
* "other" for a proprietary method defined by the solution.

**Comment by mkovatsc** [comment]
It might be better to return TDs by the Observable instead of instances of `ConsumedThing`: The script has more control over how many instances are created; `discover()` becomes the complement of the manual `fetch()`.

**Comment by mkovatsc** [comment]
What is the difference between "nearby" and "broadcast"?

"nearby" should probably talk about proximity or distance, as Bluetooth and NFC are examples for broadcast. Even rename to "proximity"?

**Comment by zolkis** [comment]
"nearby" is meant for the likes of Bluetooth and NFC. "broadcast" is for the case when a broadcast request is sent (like in OCF endpoint discovery).

Returning TDs instead of ConsumedThing instances makes a lot of sense to me.

**Comment by mkovatsc** [comment]
My point is about tweaking the terms to be more precise and intuitive given other usage. Technically Bluetooth is broadcasting (radio is always a broadcast channel).

BLE and some 802.15.4 radios support distance estimations and NFC is meant for ultimate proximity, which is why I think "proximity" might be better than "nearby", as the latter is very vague. We could also talk about an optional distance threshold parameter for the filter.

"multicast" might be the better term for the current "broadcast", as multicast discovery is the term commonly used, e.g., in CoRE and OCF.

**Comment by zolkis** [comment]
Using "broadcast" for Bluetooth would be too... broad. Then it would be more clear and useful to use `"nfc"` and `"ble"`. Also, an optional timeout would also make sense for discovery.
So the updated API would look like:

```javascript
partial interface WoT {
    Observable<ThingDescription> discover(optional ThingFilter filter);
};

typedef enum { "any", "local", "directory", "nfc", "ble", "other" } DiscoveryMethod;

dictionary ThingFilter {
    DiscoveryMethod  method = "any";
    unsigned long       timeout;
    USVString            url;
    USVString            query;  // TBD
    sequence<Dictionary> constraints; // TBD
};
```
Let's see how useful is to discover based on query (SPARQL) and constraints (depending what we do with SemanticAnnotations). `url` could be used in a number of ways, we need to be very specific how it is supposed to be used.

**Comment by mkovatsc** [comment]
> Using "broadcast" for Bluetooth would be too... broad

I have no clue who wants to do that. I proposed "proximity" as possible improvement over "nearby".

**Comment by mkovatsc** [comment]
> Let's see how useful is to discover based on query (SPARQL) and constraints 

There are a number of JSON query languages. We should look at those, e.g., https://jrgql.github.io/

**Comment by zolkis** [comment]
Fixed in #113.
